[BUG] Searchable Snapshot not working with an NPE error

**Describe the bug**
I create a snapshot based on HDFS. When the snapshot is not searchable, it works fine. When I set the snapshot to searchable, it reports error. The opensearch version is 2.9
**To Reproduce**
Create the repository:

```
{
  "type": "hdfs",
  "settings": {
    "uri": "hdfs://es1:9000",
    "path": "/searchable_snapshots",
    "conf.dfs.client.read.shortcircuit": "false"
  }
}
```
create index:
```
{
  "settings": {
    "number_of_shards": 1,
    "number_of_replicas": 0
  },
  "mappings": {
    "properties": {
      "name": {
        "type": "text"
      },
      "age": {
        "type": "integer"
      }
    }
  }
}
```

create snapshot:
```
{
  "indices": "my_index",
  "ignore_unavailable": true,
  "include_global_state": false,
  "partial": false
}
```

create restore index:
```
{
  "indices": "my_index",
  "storage_type": "remote_snapshot",
  "rename_pattern": "(.+)",
  "rename_replacement": "restored_my_index"
}
```

search. The error reports:
```
{
  "error" : {
    "root_cause" : [ ],
    "type" : "search_phase_execution_exception",
    "reason" : "all shards failed",
    "phase" : "query",
    "grouped" : true,
    "failed_shards" : [ ]
  },
  "status" : 503
}
```
Error in log:

```
[2023-08-14T14:36:06,713][WARN ][o.o.c.r.a.AllocationService] [es2] failing shard [failed shard, shard [restored_my_index][0], node[jMJDygdtQC6QqlAi043HUg], [P], recovery_source[snapshot recovery [YpS3Jwc8RROsfw7z1Wn7uQ] from searchable_hdfs_repository:searchable_snapshot/sEJNO-07T2Sc_9sO9b0UYA], s[INITIALIZING], a[id=t7IsZGhZQUKEOn1c4L5GnA], unassigned_info[[reason=ALLOCATION_FAILED], at[2023-08-14T06:36:06.588Z], failed_attempts[4], failed_nodes[[jMJDygdtQC6QqlAi043HUg]], delayed=false, details[failed shard on node [jMJDygdtQC6QqlAi043HUg]: failed recovery, failure RecoveryFailedException[[restored_my_index][0]: Recovery failed on {es2}{jMJDygdtQC6QqlAi043HUg}{leroNMlxRsOMDwkuxlyRcg}{192.168.56.103}{192.168.56.103:9300}{dms}{shard_indexing_pressure_enabled=true}]; nested: IndexShardRecoveryException[failed recovery]; nested: NullPointerException[Cannot invoke "org.opensearch.common.unit.ByteSizeValue.getBytes()" because the return value of "org.opensearch.index.snapshots.blobstore.BlobStoreIndexShardSnapshot$FileInfo.partSize()" is null]; ], allocation_status[fetching_shard_data]], message [failed recovery], failure [RecoveryFailedException[[restored_my_index][0]: Recovery failed on {es2}{jMJDygdtQC6QqlAi043HUg}{leroNMlxRsOMDwkuxlyRcg}{192.168.56.103}{192.168.56.103:9300}{dms}{shard_indexing_pressure_enabled=true}]; nested: IndexShardRecoveryException[failed recovery]; nested: NullPointerException[Cannot invoke "org.opensearch.common.unit.ByteSizeValue.getBytes()" because the return value of "org.opensearch.index.snapshots.blobstore.BlobStoreIndexShardSnapshot$FileInfo.partSize()" is null]; ], markAsStale [true]]
org.opensearch.indices.recovery.RecoveryFailedException: [restored_my_index][0]: Recovery failed on {es2}{jMJDygdtQC6QqlAi043HUg}{leroNMlxRsOMDwkuxlyRcg}{192.168.56.103}{192.168.56.103:9300}{dms}{shard_indexing_pressure_enabled=true}
        at org.opensearch.index.shard.IndexShard.lambda$executeRecovery$30(IndexShard.java:3554) ~[opensearch-2.9.0.jar:2.9.0]
        at org.opensearch.action.ActionListener$1.onFailure(ActionListener.java:88) ~[opensearch-2.9.0.jar:2.9.0]
        at org.opensearch.index.shard.StoreRecovery.lambda$recoveryListener$8(StoreRecovery.java:510) ~[opensearch-2.9.0.jar:2.9.0]
        at org.opensearch.action.ActionListener$1.onFailure(ActionListener.java:88) ~[opensearch-2.9.0.jar:2.9.0]
        at org.opensearch.action.ActionListener.completeWith(ActionListener.java:345) ~[opensearch-2.9.0.jar:2.9.0]
        at org.opensearch.index.shard.StoreRecovery.recoverFromStore(StoreRecovery.java:113) ~[opensearch-2.9.0.jar:2.9.0]
        at org.opensearch.index.shard.IndexShard.recoverFromStore(IndexShard.java:2620) ~[opensearch-2.9.0.jar:2.9.0]
        at org.opensearch.action.ActionRunnable$2.doRun(ActionRunnable.java:88) ~[opensearch-2.9.0.jar:2.9.0]
        at org.opensearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:908) ~[opensearch-2.9.0.jar:2.9.0]
        at org.opensearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:52) ~[opensearch-2.9.0.jar:2.9.0]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) [?:?]
        at java.lang.Thread.run(Thread.java:833) [?:?]
Caused by: org.opensearch.index.shard.IndexShardRecoveryException: failed recovery
        ... 11 more
Caused by: java.lang.NullPointerException: Cannot invoke "org.opensearch.common.unit.ByteSizeValue.getBytes()" because the return value of "org.opensearch.index.snapshots.blobstore.BlobStoreIndexShardSnapshot$FileInfo.partSize()" is null
        at org.opensearch.index.store.remote.file.OnDemandBlockSnapshotIndexInput.<init>(OnDemandBlockSnapshotIndexInput.java:107) ~[opensearch-2.9.0.jar:2.9.0]
        at org.opensearch.index.store.remote.file.OnDemandBlockSnapshotIndexInput.<init>(OnDemandBlockSnapshotIndexInput.java:90) ~[opensearch-2.9.0.jar:2.9.0]
        at org.opensearch.index.store.remote.file.OnDemandBlockSnapshotIndexInput.<init>(OnDemandBlockSnapshotIndexInput.java:61) ~[opensearch-2.9.0.jar:2.9.0]
        at org.opensearch.index.store.remote.directory.RemoteSnapshotDirectory.openInput(RemoteSnapshotDirectory.java:77) ~[opensearch-2.9.0.jar:2.9.0]
        at org.apache.lucene.store.FilterDirectory.openInput(FilterDirectory.java:101) ~[lucene-core-9.7.0.jar:9.7.0 ccf4b198ec328095d45d2746189dc8ca633e8bcf - 2023-06-21 11:48:16]
        at org.apache.lucene.store.FilterDirectory.openInput(FilterDirectory.java:101) ~[lucene-core-9.7.0.jar:9.7.0 ccf4b198ec328095d45d2746189dc8ca633e8bcf - 2023-06-21 11:48:16]
        at org.apache.lucene.store.Directory.openChecksumInput(Directory.java:156) ~[lucene-core-9.7.0.jar:9.7.0 ccf4b198ec328095d45d2746189dc8ca633e8bcf - 2023-06-21 11:48:16]
        at org.apache.lucene.codecs.lucene90.Lucene90CompoundReader.readEntries(Lucene90CompoundReader.java:110) ~[lucene-core-9.7.0.jar:9.7.0 ccf4b198ec328095d45d2746189dc8ca633e8bcf - 2023-06-21 11:48:16]
        at org.apache.lucene.codecs.lucene90.Lucene90CompoundReader.<init>(Lucene90CompoundReader.java:67) ~[lucene-core-9.7.0.jar:9.7.0 ccf4b198ec328095d45d2746189dc8ca633e8bcf - 2023-06-21 11:48:16]
        at org.apache.lucene.codecs.lucene90.Lucene90CompoundFormat.getCompoundReader(Lucene90CompoundFormat.java:86) ~[lucene-core-9.7.0.jar:9.7.0 ccf4b198ec328095d45d2746189dc8ca633e8bcf - 2023-06-21 11:48:16]
```


**Host/Environment (please complete the following information):**
 - OS: Centos7

**Additional context**
Add any other context about the problem here.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -162,6 +162,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 - Fix flaky ResourceAwareTasksTests.testBasicTaskResourceTracking test ([#8993](https://github.com/opensearch-project/OpenSearch/pull/8993))
 - Fix memory leak when using Zstd Dictionary ([#9403](https://github.com/opensearch-project/OpenSearch/pull/9403))
 - Fix range reads in respository-s3 ([9512](https://github.com/opensearch-project/OpenSearch/issues/9512))
+- Handle null partSize in OnDemandBlockSnapshotIndexInput ([#9291](https://github.com/opensearch-project/OpenSearch/issues/9291))
 
 ### Security
 

--- a/server/src/main/java/org/opensearch/index/store/remote/file/OnDemandBlockSnapshotIndexInput.java
+++ b/server/src/main/java/org/opensearch/index/store/remote/file/OnDemandBlockSnapshotIndexInput.java
@@ -8,8 +8,6 @@
 
 package org.opensearch.index.store.remote.file;
 
-import org.apache.logging.log4j.LogManager;
-import org.apache.logging.log4j.Logger;
 import org.apache.lucene.store.FSDirectory;
 import org.apache.lucene.store.IndexInput;
 import org.opensearch.index.snapshots.blobstore.BlobStoreIndexShardSnapshot.FileInfo;
@@ -26,8 +24,6 @@ import java.io.IOException;
  * @opensearch.internal
  */
 public class OnDemandBlockSnapshotIndexInput extends OnDemandBlockIndexInput {
-    private static final Logger logger = LogManager.getLogger(OnDemandBlockSnapshotIndexInput.class);
-
     /**
      * Where this class fetches IndexInput parts from
      */
@@ -48,7 +44,7 @@ public class OnDemandBlockSnapshotIndexInput extends OnDemandBlockIndexInput {
     protected final String fileName;
 
     /**
-     * part size  in bytes
+     * Maximum size in bytes of snapshot file parts.
      */
     protected final long partSize;
 
@@ -104,7 +100,15 @@ public class OnDemandBlockSnapshotIndexInput extends OnDemandBlockIndexInput {
         super(builder);
         this.transferManager = transferManager;
         this.fileInfo = fileInfo;
-        this.partSize = fileInfo.partSize().getBytes();
+        if (fileInfo.partSize() != null) {
+            this.partSize = fileInfo.partSize().getBytes();
+        } else {
+            // Repository implementations can define a size at which to split files
+            // into multiple objects in the repository. If partSize() is null, then
+            // no splitting happens, so default to Long.MAX_VALUE here to have the
+            // same effect. See {@code BlobStoreRepository#chunkSize()}.
+            this.partSize = Long.MAX_VALUE;
+        }
         this.fileName = fileInfo.physicalName();
         this.directory = directory;
         this.originalFileSize = fileInfo.length();
@@ -131,6 +135,10 @@ public class OnDemandBlockSnapshotIndexInput extends OnDemandBlockIndexInput {
 
         final long blockStart = getBlockStart(blockId);
         final long blockEnd = blockStart + getActualBlockSize(blockId);
+
+        // If the snapshot file is chunked, we must account for this by
+        // choosing the appropriate file part and updating the position
+        // accordingly.
         final int part = (int) (blockStart / partSize);
         final long partStart = part * partSize;
 

--- a/server/src/test/java/org/opensearch/index/store/remote/file/OnDemandBlockSnapshotIndexInputTests.java
+++ b/server/src/test/java/org/opensearch/index/store/remote/file/OnDemandBlockSnapshotIndexInputTests.java
@@ -19,6 +19,8 @@ import org.apache.lucene.store.MMapDirectory;
 import org.apache.lucene.store.SimpleFSLockFactory;
 import org.apache.lucene.util.Constants;
 import org.apache.lucene.util.Version;
+import org.opensearch.common.lucene.store.ByteArrayIndexInput;
+import org.opensearch.core.common.unit.ByteSizeUnit;
 import org.opensearch.core.common.unit.ByteSizeValue;
 import org.opensearch.index.snapshots.blobstore.BlobStoreIndexShardSnapshot;
 import org.opensearch.index.store.StoreFileMetadata;
@@ -31,9 +33,12 @@ import java.io.EOFException;
 import java.io.IOException;
 import java.nio.file.Path;
 
+import static org.mockito.ArgumentMatchers.argThat;
 import static org.mockito.Mockito.any;
 import static org.mockito.Mockito.doAnswer;
 import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
 
 @ThreadLeakFilters(filters = CleanerDaemonThreadLeakFilter.class)
 public class OnDemandBlockSnapshotIndexInputTests extends OpenSearchTestCase {
@@ -43,7 +48,6 @@ public class OnDemandBlockSnapshotIndexInputTests extends OpenSearchTestCase {
     private static final String FILE_NAME = "File_Name";
     private static final String BLOCK_FILE_PREFIX = FILE_NAME;
     private static final boolean IS_CLONE = false;
-    private static final ByteSizeValue BYTE_SIZE_VALUE = new ByteSizeValue(1L);
     private static final int FILE_SIZE = 29360128;
     private TransferManager transferManager;
     private LockFactory lockFactory;
@@ -74,7 +78,38 @@ public class OnDemandBlockSnapshotIndexInputTests extends OpenSearchTestCase {
         runAllTestsFor(22);
     }
 
-    public void runAllTestsFor(int blockSizeShift) throws Exception {
+    public void testChunkedRepository() throws IOException {
+        final long blockSize = new ByteSizeValue(1, ByteSizeUnit.KB).getBytes();
+        final long repositoryChunkSize = new ByteSizeValue(2, ByteSizeUnit.KB).getBytes();
+        final long fileSize = new ByteSizeValue(3, ByteSizeUnit.KB).getBytes();
+
+        when(transferManager.fetchBlob(any())).thenReturn(new ByteArrayIndexInput("test", new byte[(int) blockSize]));
+        try (
+            FSDirectory directory = new MMapDirectory(path, lockFactory);
+            IndexInput indexInput = new OnDemandBlockSnapshotIndexInput(
+                OnDemandBlockIndexInput.builder()
+                    .resourceDescription(RESOURCE_DESCRIPTION)
+                    .offset(BLOCK_SNAPSHOT_FILE_OFFSET)
+                    .length(FILE_SIZE)
+                    .blockSizeShift((int) (Math.log(blockSize) / Math.log(2)))
+                    .isClone(IS_CLONE),
+                new BlobStoreIndexShardSnapshot.FileInfo(
+                    FILE_NAME,
+                    new StoreFileMetadata(FILE_NAME, fileSize, "", Version.LATEST),
+                    new ByteSizeValue(repositoryChunkSize)
+                ),
+                directory,
+                transferManager
+            )
+        ) {
+            // Seek to the position past the first repository chunk
+            indexInput.seek(repositoryChunkSize);
+        }
+        // Verify the second chunk is requested (i.e. ".part1")
+        verify(transferManager).fetchBlob(argThat(request -> request.getBlobName().equals("File_Name.part1")));
+    }
+
+    private void runAllTestsFor(int blockSizeShift) throws Exception {
         final OnDemandBlockSnapshotIndexInput blockedSnapshotFile = createOnDemandBlockSnapshotIndexInput(blockSizeShift);
         final int blockSize = 1 << blockSizeShift;
         TestGroup.testGetBlock(blockedSnapshotFile, blockSize, FILE_SIZE);
@@ -106,7 +141,7 @@ public class OnDemandBlockSnapshotIndexInputTests extends OpenSearchTestCase {
         fileInfo = new BlobStoreIndexShardSnapshot.FileInfo(
             FILE_NAME,
             new StoreFileMetadata(FILE_NAME, FILE_SIZE, "", Version.LATEST),
-            BYTE_SIZE_VALUE
+            null
         );
 
         int blockSize = 1 << blockSizeShift;
@@ -182,7 +217,7 @@ public class OnDemandBlockSnapshotIndexInputTests extends OpenSearchTestCase {
 
     }
 
-    public static class TestGroup {
+    private static class TestGroup {
 
         public static void testGetBlock(OnDemandBlockSnapshotIndexInput blockedSnapshotFile, int blockSize, int fileSize) {
             // block 0

